### DISCUSSION TITLE
Problem with virtual layer

### DISCUSSION THREAD

**Discussion by theudebert** [opengisch/QField#5047] [Q&A]
I want to use qfield in ornithological field work for mapping observations. I used to do this with QGIS on a Windows Tablet, but now I want to switch to QField on an android device. For selecting a species from a long list of potentially occurring species I had written a python script, that adds at the top of the list in the drop-down menu a certain number of species that have been used most often. So for selecting a species I usually do not have to scroll through the complete list with more than 200 entries, but just tap on one of the species which appear in the visible part of the drop-down list.
The composition of the list is done via a Python script which also adds the entries of the most used species to the drop-down list.
I Hope the following screenshot from my desktop QGIS explains, what I'm talking about. Above the thin line there is a list of the most used species - easily accessible at the top of the list - and then follows the complete list.
![Dropdown_Auswahlliste](https://github.com/opengisch/QField/assets/66841475/57cb069b-3870-4e8b-8413-321e83fc1ad2)
As QField does not support Python, I tried to accomplish the same by using a virtual layer as source for the keys and values in the value relation. This works in Desktop QGIS - the list is composed correctly and occurs at the top of the drop-down.
In QField the virtual layer works correctly - it shows the list as I need it. But the point layer for the observations does not realize the change in the virtual layer and does not update the drop-down field accordingly.
I tried to set a dependency in the properties of the point layer - but this did not help.
Does anyone have an idea, how I can solve this problem?

Thank you very much for all your work

**Comment by SeqLaz** [comment]
Hello @theudebert,  
  
A better approach could be to designate a table inside a GeoPackage to register species names. Then, use a "Value Relation" widget with an auto-completer. This way, you can type, search, filter, and find values more quickly.  
  
For more information, refer to this link: https://docs.qfield.org/how-to/attributes-form/#configure-auto-complete-value
  
Additionally, in the species name table, you can designate a field with the number of observations of the species and sort the order based on the number of observations using a specific field.

**Reply by theudebert** [reply]
Well, when using a touchUI I prefer to minimize typing.

**Reply by SeqLaz** [reply]
You should give it a try using sort by the observations field

**Reply by theudebert** [reply]
1. In my virtual layer I have a field with the number of observations for every species. How could I sort the list in the drop-down by this? I think, sorting is only possible by keys or by values.

**Reply by theudebert** [reply]
2. Sorting by the number of observations does not solve my first problem. This would mean, that the drop-down list has to be updated according to dynamically created values (from the virtual layer). But this does not  work! :-(

**Comment by lp-dj** [comment]
Hi,

I was faced with the same task a few years ago and received a request from the mapping community for the most efficient input option possible.

For some time now, we have been working well with a combination of species abbreviation and leading and, if necessary, closing "escape character". I have chosen the "." here because it is usually immediately accessible on Android keyboards (the most commonly used OS in our mapping community) (to the right of the space).

I have only received positive feedback on this solution so far. All species can be accessed via a maximum of three characters, and I only need to type two characters for most of them; with a reference list of 282 species.

https://github.com/opengisch/QField/assets/80956612/b6e58aab-11cc-452f-a2b4-43898c990725

If you are interested, you can find more information here (currently under revision, preparation for the coming season): wiki.loekplan.de

Best regards
